### ISSUE TITLE
How featureful should the "featureful" branch be?

### ISSUE THREAD

**Issue by lucario387** [NvChad/example_config#21]
tl;dr: There is another branch in example_config, named "v2.0_featureful". I want to make it like how the name suggests, "featureful"
Just wondering, Idk how many plugins i should add, which i should add, and so on

Edit (5月１５日): Feel free to create PRs to add plugin supports

**Comment by bryant-the-coder** [comment]
Overseer.nvim

**Comment by siduck** [comment]
https://github.com/folke/trouble.nvim  https://github.com/simrat39/symbols-outline.nvim 

nvim-navic in statusline https://gist.github.com/siduck/8d6f892d282fb5cbdf4cf623dd50b481

**Comment by siduck** [comment]
@lucario387 do lazyloading wherever possible

**Comment by lucario387** [comment]
my currnetp lan

adding an `extras` folder into `custom.configs`, where it will just be a bunch of lazy plugin spec. Then I will tell people that if they want to use `x`, just add a line of `{ import = "custom.configs.extras.x" }` into `custom/plugins.lua`

**Comment by Daniel-Boll** [comment]
@lucario387 if you happen to need help I'm willing to contribute to the featureful branch.

Best Regards,
Daniel Boll. 🎴

**Comment by Daniel-Boll** [comment]
I mean more in the sense that you already have a plan/roadmap for the next features to be added. Perhaps we can organize tasks for each feature and work on it simultaneously.

Best Regards,
Daniel Boll. 🎴

**Comment by lucario387** [comment]
Ah no, there was never any roadmap to begin with :) 

The original question of "how featureful should it be" remains unanswered as of now. And looking ahead(somehwat) it will probably be in limbo status just like now.

Perhaps, in the short term, you can add the request overseer.nvim and navic that was requested above? I'm not familiar with any of those plugins (and most plugins released in the last year, including lazy.nvim itself)

**Comment by Daniel-Boll** [comment]
@siduck I do have access to the Project Board, but can't add nor edit anything on it.

I'm also thinking of creating a RFC in the main repo so the users may suggest and have opinion on this `featureful` and a proper discussion tab here in this repo where we can align our expectations.

Best Regards,
Daniel Boll. :flower_playing_cards:

**Comment by siduck** [comment]
> @siduck I do have access to the Project Board, but can't add nor edit anything on it.
> 
> I'm also thinking of creating a RFC in the main repo so the users may suggest and have opinion on this `featureful` and a proper discussion tab here in this repo where we can align our expectations.
> 
> Best Regards, Daniel Boll. flower_playing_cards

ok added

**Comment by Daniel-Boll** [comment]
TLDR;
Which approach to follow so the user can get the `example_config`:
- Prompt in the nvim (currently + want featureful)
- Just keep as it's know and indicate in the docs to checkout other branch

How can users _"ask"_ for features in `featureful`?
- Through `NvChad/example_config` issues?
- Through `NvChad/NvChad` discussions?
- None?

---
Currently we are cloning this repository in the `gen_chadrc_template` just by asking if they want to install the chadrc template, as we add featureful as stable we will also have to ask which template.

Any plans on how to deal with that? Consecutive questions? A prettier prompt (not entirely sure if possible as no plugins are installed yet I presume)? Or just indicate through the docs that they should checkout the other branch?

We may have to consider encouraging the user to fork `example_config` instead.

Best Regards,
Daniel Boll. :flower_playing_cards:

**Comment by siduck** [comment]
@Daniel-Boll through issues

**Comment by siduck** [comment]
i think we should have different config files for different types of development. For example nvchad for web developement, C/c++ , rust etc

**Comment by lucario387** [comment]
That just makes it so we will have to maintain multiple branches synchronously, and i don't think that's nice
Just create an extra module for that specific task instead

**Comment by siduck** [comment]
> That just makes it so we will have to maintain multiple branches synchronously, and i don't think that's nice Just create an extra module for that specific task instead

yea and import that modules in custom plugins table

**Comment by Daniel-Boll** [comment]
> Just create an extra module for that specific task instead

What do you mean?

I'm trying to picture the workflow of a user trying to have `featureful` from a fresh start.

```
     ┌─────────────┐                                                                       
     │Clones NvChad│                                                                       
     └──────┬──────┘                                                                       
  __________▽___________          ______________________     ┌────────────────────────────┐
 ╱                      ╲        ╱                      ╲    │Clones NvChad/example_config│
╱ Do you want to install ╲______╱ Do you want the        ╲___│@ main                      │
╲ chadrc template?       ╱yes   ╲ `featureful` template? ╱yes└──────────────┬─────────────┘
 ╲______________________╱        ╲______________________╱                   │              
            │no                             │no                             │              
     ┌──────▽──────┐         ┌──────────────▽─────────────┐                 │              
     │Has the basic│         │Clones NvChad/example_config │                 │              
     │chadrc file  │          │@ featureful                │                 │              
     └──────┬──────┘         └──────────────┬─────────────┘                 │              
            └─────────────────┬─────────────┴───────────────────────────────┘              
                    ┌─────────▽─────────┐                                                  
                    │Starts using NvChad│                                                  
                    └───────────────────┘                                                  
```

<p align="center">
   or
</p>

```
     ┌─────────────┐                                              
     │Clones NvChad│                                              
     └──────┬──────┘                                              
  __________▽___________     ┌────────────────────────────┐       
 ╱                      ╲    │Clones NvChad/example_config│       
╱ Do you want to install ╲___│@ main                      │       
╲ chadrc template?       ╱yes└──────────────┬─────────────┘       
 ╲______________________╱                   │                     
            │no                             │                     
     ┌──────▽──────┐                        │                     
     │Has the basic│                        │                     
     │chadrc file  │                        │                     
     └──────┬──────┘                        │                     
            └────────┬──────────────────────┘                     
           ┌─────────▽─────────┐                                  
           │Starts using NvChad│                                  
           └─────────┬─────────┘                                  
           __________▽__________     ┌───────────────────────────┐
          ╱                     ╲    │Copies manually the extras │
         ╱ User wants featureful ╲___│folder to it's `lua/custom`│
         ╲                       ╱yes└─────────────┬─────────────┘
          ╲_____________________╱                  │              
                     │no                           │              
                     └───────┬─────────────────────┘              
                    ┌────────▽────────┐                           
                    │Keep using NvChad│                           
                    └─────────────────┘                           
```

I'm not sure if I'm way off the board here, just trying to understand which kinda of experience are we trying to provide.

Best Regards,
Daniel Boll. :flower_playing_cards:

**Comment by siduck** [comment]
@Daniel-Boll check astronvim community repo. each feature i.e like plugins are in separate files and defined. so one can just use lazy's import feature to load those files

**Comment by Daniel-Boll** [comment]
Okay, so the `featureful` branch will be eventually merged into the `example_config@main` and made available generally to everyone so they can choose to import it by using lazy's import feature?

Got it

Best Regards,
Daniel Boll. :flower_playing_cards:

**Comment by Daniel-Boll** [comment]
- [ ] zbirenbaum/copilot.lua
- [ ] sindrets/diffview.nvim
- [ ] anuvyklack/windows.nvim
- [ ] simrat39/rust-tools.nvim
- [ ] p00f/clangd_extensions.nvim

Any objection for any of this plugins? 

Best Regards,
Daniel Boll. :flower_playing_cards:

**Comment by Daniel-Boll** [comment]
> 1. Optional

But isn't every plugin in extra optional already?

**Comment by lucario387** [comment]
No, optional as in im whatever, less than wanting to, adding base spec support to it.

I don't have the allowance to test copilot, and I have zero expectation the people using it will read the file and contributing when something new comes around (or at least notify the maintainers of it)

**Comment by Daniel-Boll** [comment]
Partially agree, there's actually many users of `copilot` in neovim as it's no longer a VSCode only feature, and both official `github/copilot.vim` and `zbirenbaum/copilot.lua` are actually cumbersome to configure.

Either way, AstroNvim and LazyVim both have predefined configuration for copilot, so I think is still valid.

> zero expectation the people using it will read the file and contributing when something new comes around 

If we start from these premise then nothing here will make sense :man_shrugging:, I think this can be oriented like AstroNvim as a community configured plugins.

What are your thoughts @siduck?

Best Regards,
Daniel Boll. :flower_playing_cards:

**Comment by siduck** [comment]
@Daniel-Boll i dont have copilot so i cant test it here :(

**Comment by Daniel-Boll** [comment]
I can take care of it, I do have access. My question is if you think this is a valid point or not to be considered (about the community approach).

Best Regards,
Daniel Boll. 🎴

**Comment by siduck** [comment]
yes

**Comment by siduck** [comment]
> If we are talking about astro and lazy, which also has nvim-dap, how do we "rice" it, or just delegate everything to nvim-dap-ui

leave the ricing part to me. I'll add hl groups for dap and all plugins we use in featureful branch soon

**Comment by soomtong** [comment]
As a neovim user, still using vim script based plugins.
Here's a list and a brief introduction of what I'd like to see included in nvchad full feature.

- t9md/vim-quickhl

I haven't found a replacement for this plugin yet. It simply highlights words at the cursor. Great for code reviews on site.

- rlane/pounce.nvim

A plugin in the vein of easy motion or acejump. Made with the neovim version. 

- tpope/vim-surround

I think this is an essential plugin for vim.

- mg979/vim-visual-multi

The best solution for dealing with multiple cursors.

- terryma/vim-expand-region

Works great in environments where treesitter is not supported.

- axelf4/vim-strip-trailing-whitespace

This would be a nice feature to have on top of neovim as a default.

- MattesGroeger/vim-bookmarks

A tool for managing bookmarks. It also needs a plugin to work with telescope.

- tom-anders/telescope-vim-bookmarks.nvim

A plugin to integrate the above bookmarks plugin with telescope.

- folke/persistence.nvim

Useful for when you're not using the project feature.

- f-person/git-blame.nvim

This is a feature that would be hollow without.

- Pocco81/auto-save.nvim

Has several options.

In the meantime, I'm using lunarvim with these plugins.

I have tried to apply this plugin to nvchad, but without success.
It would be very helpful if you could point me to a suitable example using the vanilla vim script plugin.

**Comment by lucario387** [comment]
Thank you very much for the request!

As for the list of plugins:
- vim-quickhl: I'd prefer to have one in lua, ig
- pounce.nvim: If I add this, there will be a question of `Why not add leap.nvim`, so maybe both?
- vim-surround: `kylechui/nvim-surround`
- vim-visual-multi: ye this is nice
- vim-expand-region: Maybe maybe not
- vim-strip-trailing-whitespace: Can't this be done with a simple autocmd? Or is it much more efficient
- bookmarks: :+1:
- persistence: Somewhat agrees
- auto-save.nvim: Doesn't the default settings work with `lazy = false` added?

**Comment by BrunoKrugel** [comment]
how should we handle mappings? add they as example in `mappings.lua`?

**Comment by lucario387** [comment]
honestly, idk how to handle mappings just yet @BrunoKrugel 

We can just hack it as 
```lua
require("core.mappings").something = {
...
}
```
for mappings
Then
```lua
require("core.utils").load_mappings("something")
```

This will work

**Comment by BrunoKrugel** [comment]
So, we should have an agreement about which plugins we can add or anyone can create a PR with a plugin that we find usefull and the users decide?

Also, alternative to `vim-quickhl` there is `RRethy/vim-illuminate`

**Comment by lucario387** [comment]
Oh feel free to create PRs. I'm not stopping that.

**Comment by siduck** [comment]
https://github.com/NvChad/nvcommunity

**Comment by HaleTom** [comment]
Does this mean that `v2.0_featureful` branch is now defuct and we should use [NvChad/nvcommunity](https://github.com/NvChad/nvcommunity) instead?

**Comment by lucario387** [comment]
yes
